File: PSET6/sentimental-readability/readability.py
from cs50 import get_string


def calculate_coleman_liau(text):
    # Count the number of letters, words, and sentences in the text
    num_letters = sum(1 for char in text if char.isalpha())
    num_words = len(text.split())
    num_sentences = text.count(".") + text.count("!") + text.count("?")

    # Calculate the average number of letters per 100 words
    L = float(num_letters) / float(num_words) * 100

    # Calculate the average number of sentences per 100 words
    S = float(num_sentences) / float(num_words) * 100

    # Calculate the Coleman-Liau index
    coleman_liau_index = 0.0588 * L - 0.296 * S - 15.8

    # Round the index to the nearest integer
    grade_level = round(coleman_liau_index)

    # Return the calculated grade level
    return grade_level


def main():
    string = get_string("Text: ")
    grade_level = calculate_coleman_liau(string)

    if grade_level < 1:
        print("Before Grade 1")
    elif grade_level >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {grade_level}")


if __name__ == '__main__':
    main()

File: PSET2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

float calculateColemanLiau(char *text);

int main(void)
{
    char *text;
    text = get_string("Text: ");
    float index = calculateColemanLiau(text);
    int colemanLiauIndex = round(index);
    if (colemanLiauIndex < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (colemanLiauIndex >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", colemanLiauIndex);
    }
}

float calculateColemanLiau(char *text)
{
    int letters = 0, words = 0, sentences = 0;

    // Iterate through each character in the text
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            // Count the number of letters in the text
            letters++;
        }
        else if (isspace(text[i]))
        {
            // Count the number of words in the text
            words++;
        }
        else if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            // Count the number of sentences in the text
            sentences++;
        }
    }

    // Increment words for the last word in the text
    words++;

    // Calculate the Coleman-Liau index
    float L = (float) letters / words * 100;
    float S = (float) sentences / words * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;

    return index;
}